lib/NetPacksLobby.h

```diff
@@ -0,0 +1,311 @@
+	bool checkClientPermissions(CVCMIServer * srv) const;
+	void applyOnLobbyScreen(CLobbyScreen * lobby, CServerHandler * handler);
+
+	template <typename Handler> void serialize(Handler &h, const int version)
+	{
+		h & action;
+	}
+};
+
+struct LobbyStartGame : public CLobbyPackToPropagate
+{
+	// Set by server
+	std::shared_ptr<StartInfo> initializedStartInfo;
+
+	LobbyStartGame() : initializedStartInfo(nullptr) {}
+	bool checkClientPermissions(CVCMIServer * srv) const;
+	bool applyOnServer(CVCMIServer * srv);
+	void applyOnServerAfterAnnounce(CVCMIServer * srv);
+	bool applyOnLobbyHandler(CServerHandler * handler);
+	void applyOnLobbyScreen(CLobbyScreen * lobby, CServerHandler * handler);
+
+	template <typename Handler> void serialize(Handler &h, const int version)
+	{
+		h & initializedStartInfo;
+	}
+};
+
+struct LobbyChangeHost : public CLobbyPackToPropagate
+{
+	int newHostConnectionId;
+
+	LobbyChangeHost() : newHostConnectionId(-1) {}
+	bool checkClientPermissions(CVCMIServer * srv) const;
+	bool applyOnServer(CVCMIServer * srv);
+	bool applyOnServerAfterAnnounce(CVCMIServer * srv);
+
+	template <typename Handler> void serialize(Handler & h, const int version)
+	{
+		h & newHostConnectionId;
+	}
+};
+
+struct LobbyUpdateState : public CLobbyPackToPropagate
+{
+	LobbyState state;
+	bool hostChanged; // Used on client-side only
+
+	LobbyUpdateState() : hostChanged(false) {}
+	bool applyOnLobbyHandler(CServerHandler * handler);
+	void applyOnLobbyScreen(CLobbyScreen * lobby, CServerHandler * handler);
+
+	template <typename Handler> void serialize(Handler &h, const int version)
+	{
+		h & state;
+	}
+};
+
+struct LobbySetMap : public CLobbyPackToServer
+{
+	std::shared_ptr<CMapInfo> mapInfo;
+	std::shared_ptr<CMapGenOptions> mapGenOpts;
+
+	LobbySetMap() : mapInfo(nullptr), mapGenOpts(nullptr) {}
+	bool applyOnServer(CVCMIServer * srv);
+
+	template <typename Handler> void serialize(Handler &h, const int version)
+	{
+		h & mapInfo;
+		h & mapGenOpts;
+	}
+};
+
+struct LobbySetCampaign : public CLobbyPackToServer
+{
+	std::shared_ptr<CCampaignState> ourCampaign;
+
+	LobbySetCampaign() {}
+	bool applyOnServer(CVCMIServer * srv);
+
+	template <typename Handler> void serialize(Handler &h, const int version)
+	{
+		h & ourCampaign;
+	}
+};
+
+struct LobbySetCampaignMap : public CLobbyPackToServer
+{
+	int mapId;
+
+	LobbySetCampaignMap() : mapId(-1) {}
+	bool applyOnServer(CVCMIServer * srv);
+
+	template <typename Handler> void serialize(Handler &h, const int version)
+	{
+		h & mapId;
+	}
+};
+
+struct LobbySetCampaignBonus : public CLobbyPackToServer
+{
+	int bonusId;
+
+	LobbySetCampaignBonus() : bonusId(-1) {}
+	bool applyOnServer(CVCMIServer * srv);
+
+	template <typename Handler> void serialize(Handler &h, const int version)
+	{
+		h & bonusId;
+	}
+};
+
+struct LobbyChangePlayerOption : public CLobbyPackToServer
+{
+	enum EWhat : ui8 {UNKNOWN, TOWN, HERO, BONUS};
+	ui8 what;
+	si8 direction; //-1 or +1
+	PlayerColor color;
```

Comment:
Passing "each click" to server is not the best idea, pass selected value (with possible delay so if client change value fast only last is propogated).

Reply:
I have some ideas how to improve this by verifying state on "host" player so changes for that player UI can be immediately applied for best responsiveness while server would still actually control everything.

But it's minor thing that I probably won't work on yet. After all with local server it's applied almost immediately and for multiplayer we better to get some actually working lobby rather than delay landing this code.